merlion/models/utils/rolling_window_dataset.py

```diff
@@ -0,0 +1,192 @@
+                return past_ts, future_ts
+            else:
+                past = self.data[idx: idx_end]
+                future = self.target[idx_end: idx_end + self.forecast_steps]
+                future_timestamp = self.target_timestamp[idx_end: idx_end + self.forecast_steps]
+                return past, future, future_timestamp
+        elif self.label_axis == 1:
+            if self.ts_index:
+                past_ts = self.data_ts[idx: idx_end]
+                future_ts = self.data_ts[idx_end]
+                return past_ts, future_ts
+            else:
+                past = self.data[idx: idx_end]
+                future = self.data[idx_end]
+                future_timestamp = np.atleast_1d(self.target_timestamp[idx_end])
+                return past, future, future_timestamp
+
+    def _get_entire_train_data_along_time(self):
+        """
+        default rolling window processor for the model to consume data as the (inputs, labels), so it gives out
+        train and label on a rolling window basis, in the format of numpy array
+        return shape:
+                inputs.shape = [n_samples, n_seq * maxlags]
+                labels.shape = [n_samples, forecast_steps]
+        """
+        inputs = np.zeros((self.valid_rolling_steps + 1, self.maxlags * self.dim))
+        for i in range(self.maxlags, len(self.data) - self.forecast_steps + 1):
+            inputs[i - self.maxlags] = self.data[i - self.maxlags: i].reshape(-1, order="F")
+
+        labels = np.zeros((self.valid_rolling_steps + 1, self.forecast_steps))
+        for i in range(self.maxlags, len(self.data) - self.forecast_steps + 1):
+            labels[i - self.maxlags] = self.target[i: i + self.forecast_steps]
+
+        labels_timestamp = self.target_timestamp[self.maxlags: len(self.data) - self.forecast_steps + 1]
+
+        return inputs, labels, labels_timestamp
+
+    def _get_entire_train_data_along_sequence(self):
+        """
+        regressive window processor for the auto-regression model to consume data as the (inputs, labels),
+        so it gives out train and label on a rolling window basis auto-regressively, in the format of numpy array
+        return shape:
+                inputs.shape = [n_samples, n_seq * maxlags]
+                labels.shape = [n_samples, n_seq]
+        """
+
+        inputs = np.zeros((len(self.data) - self.maxlags, self.maxlags * self.dim))
+        labels = np.zeros((len(self.data) - self.maxlags, self.dim))
+
+        for i in range(self.maxlags, len(self.data)):
+            inputs[i - self.maxlags] = self.data[i - self.maxlags: i].reshape(-1, order="F")
+            labels[i - self.maxlags] = self.data[i]
+
+        labels_timestamp = self.target_timestamp[self.maxlags: len(self.data)]
+
+        return inputs, labels, labels_timestamp
+
+
+def max_feasible_forecast_steps(data: TimeSeries, maxlags: int):
+    return len(data) - maxlags
```

Comment:
How necessary is this helper function? Maybe you could just check it in the initializer of the rolling window dataset?

Reply:
this is a helper function inside some model class